### PR TITLE
Move Hashtable and Set primitives to RzUtil

### LINKED ISSUE #3894: Move Hashtable and Set primitives from SDB to RzUtil
Currently they are located in these files:
- `librz/util/sdb/src/ht_*.{c,h}`
- `librz/util/sdb/src/set.{c,h}`

They should be moved upward, to `librz/util/` and properly documented with Doxygen.

### PR DESCRIPTION
**Pull request by vanshjangir**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Move Hashtable from `librz/util/sdb/src/ht_*.c` to `librz/util/ht` and 
`librz/util/sdb/src/ht_*.h` to `librz/include/rz_util` and
Set primitives from `librz/util/sdb/src/set.{c,h}` to `librz/util/`.
Update `librz/util/meson.build` , `librz/util/sdb/src/meson.build` and `librz/include/meson.build` for changed hashtables and set primitives file paths.

**Test plan**

...

**Closing issues**

Closes https://github.com/rizinorg/rizin/issues/3894

### CODE REVIEW

File: librz/util/meson.build [hunk 1]
```diff
@@ -112,7 +117,7 @@ else
   rz_util_sources += ['big.c']
 endif
 
-rz_util_includes = [platform_inc, include_directories('sdb/src', 'sdb')]
+rz_util_includes = [platform_inc, include_directories('sdb/src', 'sdb', '.')]
```
Comment (ret2libc) [inline librz/util/meson.build]: Is this really necessary?

**Comment by ret2libc** [comment]
I thought we would have encountered some issues in the moving, but it seems all is fine, nice :D

File: librz/util/meson.build (hunk 1 above)
Comment (vanshjangir) [reply librz/util/meson.build]: without this ninja -C build test is failing. I guess the failure is because the path to hashtable ht_*.h is not specified. So I added '.'

**Comment by XVilka** [comment]
I wonder if it makes sense to put them into a subdirectory since there are many files? Maybe something like `librz/util/ht/`?

**Comment by vanshjangir** [comment]
Should I move only hastables to a new subdirectory `librz/util/ht` or move both hashtables and set primitives

**Review by XVilka — CHANGES_REQUESTED** [review]
Better to move header files to `librz/include/rz_util/` though.

File: librz/include/meson.build [hunk 2]
```diff
@@ -117,6 +117,7 @@ rz_util_files = [
   'rz_util/rz_utf8.h',
   'rz_util/rz_version.h',
   'rz_util/rz_x509.h',
+
```
Comment (XVilka) [inline librz/include/meson.build]: Unnecessary change

Comment (XVilka) [reply librz/include/meson.build]: But you need to install them too in fact.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/include/meson.build (hunk 2 above)
Comment (vanshjangir) [reply librz/include/meson.build]: Do I need to install the hashtables in `librz/include/meson.build`. Because the hashtables path is present in `librz/util/meson.build`

Comment (XVilka) [reply librz/include/meson.build]: After you move includes with exporting APIs to the `librz/include/rz_util`, of course.

File: librz/util/meson.build [hunk 3]
```diff
@@ -112,7 +117,7 @@ else
   rz_util_sources += ['big.c']
 endif
 
-rz_util_includes = [platform_inc, include_directories('sdb/src', 'sdb')]
+rz_util_includes = [platform_inc, include_directories('sdb/src', 'sdb', '../include/rz_util', '.')]
```
Comment (ret2libc) [inline librz/util/meson.build:120]: I do not think these additions are needed. Why did you add them?

Comment (vanshjangir) [reply librz/util/meson.build:120]: when I was testing (ninja -C build test), some error were there, showing
`FAILED: librz/util/librz_util.so.0.7.0.p/astr.c.o` and 
`/librz/include/rz_list.h:5:10: fatal error: ht_pp.h: No such file or directory`
    `5 | #include <ht_pp.h>`
    
so i thought adding `'.'` and `../include/rz_util` will specify the path to look for the header files (ht_*.h and set.h)

**Comment by wargio** [comment]
i'm ok with these changes, but tbh, i think the original headers should be internals and we should provide explicit function headers somewhere.

@XVilka @ret2libc @thestr4ng3r @kazarmy @Rot127

**Comment by vanshjangir** [comment]
@wargio Can you describe in detail, if there is something left to change.

**Comment by wargio** [comment]
> @wargio Can you describe in detail, if there is something left to change.

before you do this, i would prefer to see if the others are also ok with my suggestion. sorry. i should have had explicitly said this to avoid you to do extra useless work :/

**Comment by ret2libc** [comment]
> i'm ok with these changes, but tbh, i think the original headers should be internals and we should provide explicit function headers somewhere.
> 
> @XVilka @ret2libc @thestr4ng3r @kazarmy @Rot127

Do you mean instead of having a generic ht_inc.h, just "unroll" the various macros and provide a regular .h? I was thinking about that, but I'm unsure... also, I think it is ok to merge as-is anyway, because it is no worse than what we have right now (those headers are just under the sdb directory right now).

**Comment by XVilka** [comment]
> > i'm ok with these changes, but tbh, i think the original headers should be internals and we should provide explicit function headers somewhere.
> > @XVilka @ret2libc @thestr4ng3r @kazarmy @Rot127
> 
> Do you mean instead of having a generic ht_inc.h, just "unroll" the various macros and provide a regular .h? I was thinking about that, but I'm unsure... also, I think it is ok to merge as-is anyway, because it is no worse than what we have right now (those headers are just under the sdb directory right now).

I agree with @ret2libc, let's merge this as is first. Having just one HT header would be nice but it's a separate thing and can be done later.

**Comment by vanshjangir** [comment]
The changes requested are done, so I thought I might ask if there is anything left to do.

**Review by XVilka — APPROVED** [review]
LGTM after that Include path problem is solved.

**Comment by wargio** [comment]
> > i'm ok with these changes, but tbh, i think the original headers should be internals and we should provide explicit function headers somewhere.
> > @XVilka @ret2libc @thestr4ng3r @kazarmy @Rot127
> 
> Do you mean instead of having a generic ht_inc.h, just "unroll" the various macros and provide a regular .h? I was thinking about that, but I'm unsure... also, I think it is ok to merge as-is anyway, because it is no worse than what we have right now (those headers are just under the sdb directory right now).

exactly.

**Review by ret2libc — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]

**Comment by XVilka** [comment]
@vanshjangir could you please fix also this error during the rz-ghidra compilation?

https://github.com/rizinorg/rz-ghidra/issues/332

**Comment by vanshjangir** [comment]
> @vanshjangir could you please fix also this error during the rz-ghidra compilation?
> 
> [rizinorg/rz-ghidra#332](https://github.com/rizinorg/rz-ghidra/issues/332)

@XVilka sure, I will fix that
